### PR TITLE
Add "Collection from..." in Checkout sidebar when selecting local pickup

### LINKED ISSUE #7997: Local Pickup: show "Pickup from [location address]" when local pickup is selected.
When you select a shipping option, you get a "shipping to [customer address]" in Cart and Checkout sidebars. When you select Local Pickup, we hide that section completely.

Instead, we can show "Pickup from [location address]" to signal the actual place the order will be in.

### PR DESCRIPTION
**Pull request by opr**
This PR is based on `add/local-pickup-methods` which has a PR here: https://github.com/woocommerce/woocommerce-blocks/pull/8256

In this PR I made the following changes:
- Add a `PickupLocation` component. This will render when the user has chosen "Local pickup" as their shipping method.
- Move `areRatesCollectible` function outside of the `useShippingData` hook. It made more sense not to create new functions every time the hook is called. For reference, this function which will check whether the selected rate is a collectible one (collectible methods are saved in the site setting `collectibleMethodIds`) - this is required to know whether we should be dispatching the rate change to all packages (if collectible, yes) or just to a single package (if not collectible this is OK).
- Update the `wc/store/checkout` data store's `prefersCollection` selector to check all collectible methods, not just `local_pickup`
- Add tests for `ShippingAddress` component, but only ones that cover this PR, new tests should be added to cover some address scenarios (these could be added within `ShippingLocation` actually)
- Add tests for `PickupLocation`
- Ensure only the local pickup method name appears in the shipping totals if using multiple packages

Fixes #7997

### Screenshots

| Before (Note the multiple package names & no pickup address) | After |
| ------ | ----- |
| image | image |

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### Internal developer Testing

<details>
<summary>Expand for example shipping method</summary>

```php
function amazon_locker_shipping_init() {
	class Amazon_Locker_Shipping_Method extends WC_Shipping_Method {

		/**
		 * Min amount to be valid.
		 *
		 * @var integer
		 */
		public $min_amount = 0;

		/**
		 * Requires option.
		 *
		 * @var string
		 */
		public $requires = '';

		/**
		 * Constructor.
		 *
		 * @param int $instance_id Shipping method instance.
		 */
		public function __construct( $instance_id = 0 ) {
			$this->id                 = 'amazon_locker_shipping';
			$this->instance_id        = absint( $instance_id );
                        $this->title              = 'Amazon locker';
			$this->method_title       = __( 'Amazon locker', 'woocommerce' );
			$this->method_description = __( 'Get your order shipped to an amazon locker.', 'woocommerce' );
			$this->supports           = array(
				'instance-settings',
				'instance-settings-modal',
				'local-pickup',
			);

			$this->init();
		}

		/**
		 * Initialize Amazon Locker shipping.
		 */
		public function init() {
		}

		/**
		 * See if Amazon locker shipping is available based on the package and cart.
		 *
		 * @param array $package Shipping package.
		 * @return bool
		 */
		public function is_available( $package ) {
			return true;
		}

		/**
		 * Called to calculate shipping rates for this method. Rates can be added using the add_rate() method.
		 *
		 * @param array $package Shipping package.
		 * @uses WC_Shipping_Method::add_rate()
		 */
		public function calculate_shipping( $package = array() ) {
			$this->add_rate(
				array(
					'label'   => $this->title,
					'cost'    => 0,
					'taxes'   => false,
					'package' => $package,
				)
			);
		}
	}
}

add_action( 'woocommerce_shipping_init', 'amazon_locker_shipping_init' ); // use this hook to initialize your new custom method

function add_amazon_locker_shipping( $methods ) {
	$methods['amazon_locker_shipping'] = 'Amazon_Locker_Shipping_Method';

	return $methods;
}
add_filter( 'woocommerce_shipping_methods', 'add_amazon_locker_shipping' );
```
</details>

2. Enable this method for one of your shipping zones. Note, the settings UI doesn't work properly for this method but that is OK.
3. Enable the built-in local pickup too (WooCommerce -> Settings -> Shipping -> Local pickup) and add a location. Please ensure you set an address for this location. **Set up two of these!**
4. Enable some regular methods (flat rate, free) for your zone too.
5. Enable the [Multiple Packages for WooCommerce](https://wordpress.org/plugins/multiple-packages-for-woocommerce/) plugin, go to its settings page (WooCommerce -> Settings -> Multiple packages) and change the `Group By` option to be `Product (individual)`.
6. Add some items to your cart (at least 2) go to the Checkout page.
7. For the `Shipping method`, select `Shipping` and then select a different rate for each package. It should look like this.
image
8. In the sidebar, ensure you see the shipping total correctly and that it lists both methods for shipping like so:
image
9. Change shipping method back to `Local Pickup` and ensure that the sidebar updates to only show **one** shipping method name.
10. Select the Amazon locker option and check the sidebar, only one method name should show, and it should also say "Collection from Amazon locker"
11. Choose one of the built-in local pickup rates you set up in step 3.
12. Ensure the sidebar updates and shows the correct method name and shows "Collection from <address>"

### User facing testing

1. Enable the built-in local pickup method (WooCommerce -> Settings -> Shipping -> Local pickup) and add two locations. Please ensure you set a different address for both of these locations.
2. Enable some different regular shipping methods (flat rate, free) for your zone too.
3. Add at least two items to your cart and go to the Checkout page.
4. For the `Shipping Method` option, select "Shipping"
5. Fill in your address and ensure the rates you set up in step 2 are shown. Select one.
6. In the Checkout sidebar, ensure you see the correct rate, and your entered address.
7. Return to the `Shipping Method` option, select "Local pickup"
8. Check the sidebar and ensure the sidebar shows local pickup, and shows the address for the pickup location you have chosen.
9. Change the location and ensure the sidebar updates.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Show the collection address in the shipping section of the Checkout sidebar when using a Local Pickup method.

### CODE REVIEW

**Review by senadir — DISMISSED** [review]
Thank you for working on this Thomas!

Unfortunately the current iteration still includes the `(change address)` text and it's confusing because we're showing the local pickup address, which is not something you can change.

image

I'd refrain from reviewing the code because I know removing `(change address)` would switch up and complicate the code a bit more.

**Comment by opr** [comment]
Hey @senadir thanks for the review and sorry for the delay getting it updated.

I fixed it up now so the conflicts are resolved and the change address button is hidden when a local pickup rate is chosen.

I will ask someone else to review since you're on rotation.

**Comment by wavvves** [comment]
Maybe I messed up testing notes but Amazon locker does not show on Local Pickup:

image

It does show as a regular shipping method, but since it's not pickup it won't appear as "collection from.." on the sidebar

image

Every step worked up until step 10

**Comment by opr** [comment]
@wavvves thanks for your patience

> Maybe I messed up testing notes but Amazon locker does not show on Local Pickup:

No, you didn't. The shipping method declared incorrect compatibility, I fixed it now!

I also improved the `useSelect` mock in the unit tests in [4e6ad75](https://github.com/woocommerce/woocommerce-blocks/pull/8305/commits/4e6ad75046992b3089822bcdee39db3041498fd2). Previously the mock returned `{ prefersCollection: false }` for all invocations, this one is more selective and prevents components who rely on `useSelect` further down the tree from breaking.

**Comment by wavvves** [comment]
Ok, so I tested everything now, and it works as expected! I did run into another problem, Amazon locker appears twice for me, but I looked into that a bit, and it seems to be related to instance_id, causing it to return two entries from the API.

image


I'm approving this, assuming it's the testing snippet's fault 👍🏼

**Review by wavvves — APPROVED** [review]
Ok, so I tested everything now, and it works as expected! I did run into another problem, Amazon locker appears twice for me, but I looked into that a bit, and it seems to be related to instance_id, causing it to return two entries from the API.

image


I'm approving this, assuming it's the testing snippet's fault 👍🏼

**Comment by opr** [comment]
@wavvves it appears twice for you on the front end or just the API response? Can you show a screenshot of that if it's on the front end please?

**Comment by wavvves** [comment]
> @wavvves it appears twice for you on the front end or just the API response? Can you show a screenshot of that if it's on the front end please?

image

(the price doesn't show as FREE because I tweaked around)

Still, it is rendering the API response correctly, which duplicates the shipping method

**Comment by opr** [comment]
Ok that's not supposed to happen, let me look into it tomorrow! Please remove your approval if you can

**Comment by opr** [comment]
@wavvves - I copy pasted the testing snippet exactly and only see the locker once. 

Could you set up a clean site and see if it happens there?

Cc @nielslange do you think you could take a look at this too and see if you get the duplicate pickup method?

**Comment by nielslange** [comment]
> Cc @nielslange do you think you could take a look at this too and see if you get the duplicate pickup method?

@opr On my end, the `Amazon locker` local pickup method only appears once, both on the Checkout block and in the API response.

Screenshot 2023-04-05 at 12 51 47

cc: @wavvves

**Review by nielslange — APPROVED** [review]
Pre-approving this, as the PR worked as expected on my end. But let's wait for @wavvves's reply to https://github.com/woocommerce/woocommerce-blocks/pull/8305/#issuecomment-1490925822, before merging it.

**Review by wavvves — APPROVED** [review]
Installed on a clean copy, everything's ok now 👍🏼

**Pull request merged**
